### PR TITLE
Upgradeable Yield Proxy

### PR DESCRIPTION
**Pull request by gakonst**
This PR separates the authorization flow of the proxy from the actual proxy implementation.

Users will always authorize 1 version of the proxy via the `onboard` and `authorizePool` calls, and then any subsequent calls will be forwarded to the address stored at `IMPL_SLOT`, which is set by an admin that's stored at the `ADMIN_SLOT` storage slots. These are made to be EIP1967 compatible to avoid storage collisions.

TODO:
- Test that upgrading to another proxy does not require re-authorizing.

### CODE REVIEW

File: contracts/peripheral/YieldProxy.sol [hunk 1]
```diff
@@ -158,394 +52,71 @@ contract YieldProxy is DecimalMath {
...
-        );
-        fyDaiObtained = fyDaiObtained.add(fyDaiBought);
-        
-        uint256 fyDaiUsed;
-        if (fyDaiObtained > 0 && controller.debtFYDai(CHAI, maturity, msg.sender) > 0) {
-            fyDaiUsed = controller.repayFYDai(CHAI, maturity, address(this), msg.sender, fyDaiObtained);
-        }
-        uint256 fyDaiRemaining = fyDaiObtained.sub(fyDaiUsed);
+// proxy contract with no storage
+contract YieldProxy is YieldAuth {
```
Comment (albertocuestacanada) [inline contracts/peripheral/YieldProxy.sol:58]: This contract seems to have the upgradable functions, but doesn't inherit from ´ProxyStorage.sol`. I'm confused.

**Comment by gakonst** [comment]
The `YieldAuth` contract is only responsible for batching `permit` / `addDelegateBySignature` calls together.

The `YieldProxy` contract inherits `YieldAuth` and implements the [Unstructured Storage](https://blog.openzeppelin.com/upgradeability-using-unstructured-storage/) pattern. It forwards all calls to the actual implementation via `delegatecall`. This means that even though the contract has no explicit state variables (since it does not inherit `ProxyStorage`), the delegatecall makes it so that: `YieldProxy` calls `ProxyV1`, adjusting `YieldProxy`'s storage, instead of `ProxyV1`'s.

The `ProxyStorage` contract exists so that we can do "safe" upgrades by only appending storage slots. If you'd want to write ProxyV2 with new state variables, you'd create a new ProxyStorageV2 and you'd inherit from `. I do that in the mocks, and you can also see how Compound does this [here](https://github.com/compound-finance/compound-protocol/blob/master/contracts/ComptrollerStorage.sol#L131).

The deployment process is:

1. Deploy the `implementation`, `ProxyV1.sol`
2. Deploy the `proxy`, `YieldProxy.sol`
3. Call `proxy.upgradeTo(implementation.address, <abi encoded initializer args>)`
4. Use `proxy.address` in the UI.

ProxyV1 is never exposed directly to the user. The user only permit/addDelegate's YieldProxy. Does that make things clearer?

File: contracts/peripheral/YieldProxy.sol [hunk 2]
```diff
@@ -158,394 +52,71 @@ contract YieldProxy is DecimalMath {
...
-        (uint256 daiObtained, uint256 fyDaiObtained) = pool.burn(msg.sender, address(this), poolTokens);
-        if (fyDaiObtained > 0) {
-            daiObtained = daiObtained.add(fyDai.redeem(address(this), address(this), fyDaiObtained));
-        }
-        
-        // Repay debt
-        if (daiObtained > 0 && controller.debtFYDai(CHAI, maturity, msg.sender) > 0) {
-            controller.repayDai(CHAI, maturity, address(this), msg.sender, daiObtained);
+        // change it
+        slot = IMPL_SLOT;
```
[the remark this answers is not in the document]
Comment (gakonst) [reply contracts/peripheral/YieldProxy.sol:83]: Yes that's the reason, I don't mind doing it although seems redundant

File: contracts/peripheral/YieldProxy.sol (hunk 1 above)
Comment (gakonst) [reply contracts/peripheral/YieldProxy.sol:58]: Explained below

**Comment by albertocuestacanada** [comment]
That makes things a lot clearer

**Comment by albertocuestacanada** [comment]
One thing that we need to implement from the very first PR is a timelock.

The first deployment will be fine, because users will have to authorized the upgradable proxy. From that time onwards, though, we could upgrade the contract in a way that allows us to steal all of our users' funds.

To avoid this, there should be at least a timelock on upgrades, and we should be scrupulous about how we communicate them.

**Comment by albertocuestacanada** [comment]
I think we are ready to merge this. I still have it grabbed by the fingertips, but I can develop new proxies and manage versioning and upgrades.

**Comment by albertocuestacanada** [comment]
Closed due to unsafe design.

**Pull request closed without merging**